File: web/src/features/messaging/api/messagingApi.js
import axiosInstance from '../../../shared/utils/api';
import { API_ENDPOINTS } from '../../../shared/utils/constants';

const unwrap = (response) => response.data?.data || response.data;

const withMessagingError = async (request) => {
  try {
    return unwrap(await request());
  } catch (error) {
    throw {
      status: error.response?.status,
      message: error.response?.data?.error
        || error.response?.data?.message
        || (error.response?.status === 403 ? 'Messaging is only available to registered users and agents.' : null)
        || error.message
        || 'Messaging request failed',
    };
  }
};

export const createConversation = async (payload) => {
  return withMessagingError(() => axiosInstance.post(API_ENDPOINTS.MESSAGING.CONVERSATIONS, payload));
};

export const getConversations = async () => {
  return withMessagingError(() => axiosInstance.get(API_ENDPOINTS.MESSAGING.CONVERSATIONS));
};

export const getMessages = async (conversationId) => {
  return withMessagingError(() => axiosInstance.get(API_ENDPOINTS.MESSAGING.MESSAGES(conversationId)));
};

export const sendMessage = async (conversationId, payload) => {
  return withMessagingError(() => axiosInstance.post(API_ENDPOINTS.MESSAGING.MESSAGES(conversationId), payload));
};

export const markConversationRead = async (conversationId) => {
  return withMessagingError(() => axiosInstance.put(API_ENDPOINTS.MESSAGING.READ(conversationId)));
};

export default {
  createConversation,
  getConversations,
  getMessages,
  sendMessage,
  markConversationRead,
};

File: mobile/app/src/main/java/com/example/fortpointproperties/features/messaging/network/MessagingApi.kt
package com.example.fortpointproperties.features.messaging.network

import com.example.fortpointproperties.features.messaging.data.model.ConversationDto
import com.example.fortpointproperties.features.messaging.data.model.CreateConversationRequest
import com.example.fortpointproperties.features.messaging.data.model.MessageDto
import com.example.fortpointproperties.features.messaging.data.model.SendMessageRequest
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.PUT

interface MessagingApi {
    @POST("api/messaging/conversations")
    suspend fun createConversation(
        @Body request: CreateConversationRequest,
    ): Response<ConversationDto>

    @GET("api/messaging/conversations")
    suspend fun getConversations(): Response<List<ConversationDto>>

    @GET("api/messaging/conversations/{conversationId}/messages")
    suspend fun getMessages(
        @Path("conversationId") conversationId: Long,
    ): Response<List<MessageDto>>

    @PUT("api/messaging/conversations/{conversationId}/read")
    suspend fun markConversationRead(
        @Path("conversationId") conversationId: Long,
    ): Response<ConversationDto>

    @POST("api/messaging/conversations/{conversationId}/messages")
    suspend fun sendMessage(
        @Path("conversationId") conversationId: Long,
        @Body request: SendMessageRequest,
    ): Response<MessageDto>
}
